src/Features/JsonPatch.SystemTextJson/src/Internal/ConversionResultProvider.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+
+namespace Microsoft.AspNetCore.JsonPatch.SystemTextJson.Internal;
+
+internal static class ConversionResultProvider
+{
+    internal static ConversionResult ConvertTo(object value, Type typeToConvertTo, JsonSerializerOptions serializerOptions)
+    {
+        if (value == null)
+        {
+            return new ConversionResult(IsNullableType(typeToConvertTo), null);
+        }
+
+        if (typeToConvertTo.IsAssignableFrom(value.GetType()))
+        {
+            // No need to convert
+            return new ConversionResult(true, value);
+        }
+
+        // Workaround for the https://github.com/dotnet/runtime/issues/113926
+        if (typeToConvertTo.Name == "JsonValuePrimitive`1")
+        {
+            typeToConvertTo = typeof(JsonNode);
+        }
+
+        try
+        {
+            var deserialized = ConvertToTargetType(value, typeToConvertTo, serializerOptions);
+            return new ConversionResult(true, deserialized);
+        }
+        catch
+        {
+            return new ConversionResult(canBeConverted: false, convertedInstance: null);
+        }
+    }
+
+    internal static ConversionResult CopyTo(object value, Type typeToConvertTo, JsonSerializerOptions serializerOptions)
+    {
+        var targetType = typeToConvertTo;
+        if (value == null)
+        {
+            return new ConversionResult(canBeConverted: true, convertedInstance: null);
+        }
+
+        if (typeToConvertTo != value.GetType() && typeToConvertTo.IsAssignableFrom(value.GetType()))
```

Comment:
This feels wrong because it is using `GetType` instead of information from `JsonTypeInfo` or `serializerOptions.GetConverter`. I think the input to this method should really be the type of the destination (the destination property's type), the type of source (the source property's type) and the  value. And the copy should serialize the value as the source type and then deserialize as the destination type. One benefit of this is that the server (or whoever is doing the patching of the CLR objects) can decide which types allow polymorphism (by using `JsonDerivedTypeAttribute`) which prevents unintentionally leaking internal data.

Reply:
Discussed this with Pranav, and I'm going to leave this as is. This would change the behavior in comparison with the original implementation, so keeping it consistent.